[WIP] Add tox and tox-conda configuration for test execution.

asford opened the pull request:
Convert existing travis test setup into a [tox](https://tox.readthedocs.io/en/latest/) and [tox-conda](https://github.com/tox-dev/tox-conda) based configuration, enabling local test execution. Add python 3.5-3.7 tox environments.

- [ ] Setup tox "dev" env for interactive development.
- [ ] Setup tox linting environment.
- [ ] Convert travis configuration to invoke tox. Should likely update to [tox-travis](https://github.com/tox-dev/tox-travis). 
- [x] Update tests to gracefully skip if docker is not available.
- [ ] Update tests to gracefully skip if fuse is not available. (Currently freezes in fuse-based tests.)

asford commented:
@martindurant This pull captures the configuration I've use to locally reproduce the test environment. Between this pull & the blackening I'm know I'm asking a number of workflow changes. Would moving to tox for test environments be workable for you?

asford commented:
Yikes, this update reveals a few py35 incompatibilities I unintentionally introduced in the test layer in #130.

asford commented:
This pull captures some work I've done to get the project's tests running locally in my environment. There are a few outstanding issues:

1) There are some py35 test failures related to py36-based features in the zipfile filesystem.
2) I'm seeing a strange failure in the fuse tests, in which the process deadlocks after receiving an ioctl call in the fuse interface. 

As these are pre-existing failures on py35, could you review this PR to see if the tox-based test and travis configuration works for you?

martindurant commented:
Happy to skip those compressions on py35. Does fuse work for you locally? The travis instances are slow and anything even remotely race-y can fail there that you would never expect to see elsewhere. I haven't come across it.

asford commented:
Sorry, this was unclear in my notes. The fuse test environment is failing locally for me, leaving a hung process. If I then `sudo umount --force` the target directory the test will unblock and fail.

It *looks* like this may be due to https://bugs.python.org/issue34070, as the ioctl call (fuse.py:116) is getting arg 21505 (0x5401, TCGETS). In this trace I then force unmount and let the test fail.

As far as I can tell, the fuse test also fails on travis with a deadlock. I believe that current master is installing fusepy into the base conda environment and skipping the fuse tests. 

```
========================================================================== FAILURES ==========================================================================
_________________________________________________________________________ test_basic _________________________________________________________________________

tmpdir = '/tmp/pytest-of-lexaf/pytest-6/test_basic0'

    def test_basic(tmpdir):
        tmpdir = str(tmpdir)
        fs = MemoryFileSystem()
        fs.touch('/mounted/testfile')
        logger.debug(fs)

        th = run(fs, '/mounted/', tmpdir, False)
        logger.debug("run %s", tmpdir)

        timeout = 10
        while True:
            try:
                logger.debug("listdir")
                # can fail with device not ready while waiting for fuse
                if 'testfile' in os.listdir(tmpdir):
                    logger.debug("break")
                    break
            except:
                pass
            timeout -= 1
            time.sleep(1)
            assert timeout > 0, "Timeout"

        logger.debug("write")
        fn = os.path.join(tmpdir, 'test')
        with open(fn, 'wb') as f:
            f.write(b'data')
        logger.debug("info")
        assert fs.info("/mounted/test")['size'] == 4

        logger.debug("open")
        f = open(fn)

        logger.debug("read")
>       assert f.read() == "data"
E       OSError: [Errno 107] Transport endpoint is not connected

fsspec/tests/test_fuse.py:46: OSError
--------------------------------------------------------------------- Captured log call ----------------------------------------------------------------------
DEBUG    fsspec.tests.test_fuse:test_fuse.py:16 <fsspec.implementations.memory.MemoryFileSystem object at 0x7f7a1605f950>
TRACE    fsspec.fuse.FUSEr:fuse.py:14 CALL *(<fsspec.implementations.memory.MemoryFileSystem object at 0x7f7a1605f950>, '/mounted/') **{}
DEBUG    fsspec.tests.test_fuse:test_fuse.py:19 run /tmp/pytest-of-lexaf/pytest-6/test_basic0
TRACE    fsspec.fuse.FUSEr:fuse.py:14 RETURN None
DEBUG    fsspec.tests.test_fuse:test_fuse.py:24 listdir
DEBUG    fsspec.tests.test_fuse:test_fuse.py:24 listdir
TRACE    fsspec.fuse.FUSEr:fuse.py:20 CALL *('/', None) **{}
TRACE    fsspec.fuse.FUSEr:fuse.py:20 RETURN {'st_uid': 1000, 'st_gid': 1000, 'st_mode': 16895, 'st_size': 0, 'st_blksize': 0, 'st_atime': 1568647084.7973132, 'st_ctime': 1568647084.797317, 'st_mtime': 1568647084.7973173}
TRACE    fsspec.fuse.FUSEr:fuse.py:43 CALL *('/', 0) **{}
TRACE    fsspec.fuse.FUSEr:fuse.py:43 RETURN ['.', '..', 'testfile']
DEBUG    fsspec.tests.test_fuse:test_fuse.py:27 break
DEBUG    fsspec.tests.test_fuse:test_fuse.py:35 write
TRACE    fsspec.fuse.FUSEr:fuse.py:20 CALL *('/test', None) **{}
DEBUG    fuse:fuse.py:740 FUSE operation getattr raised a <class 'fuse.FuseOSError'>, returning errno 2.
Traceback (most recent call last):
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/fsspec/fuse.py", line 23, in getattr
    info = self.fs.info(path)
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/fsspec/spec.py", line 509, in info
    raise FileNotFoundError(path)
FileNotFoundError: /mounted/test

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/fuse.py", line 734, in _wrapper
    return func(*args, **kwargs) or 0
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/fuse.py", line 774, in getattr
    return self.fgetattr(path, buf, None)
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/fuse.py", line 1027, in fgetattr
    attrs = self.operations('getattr', self._decode_optional_path(path), fh)
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/fuse.py", line 1076, in __call__
    return getattr(self, op)(*args)
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/autologging.py", line 1039, in autologging_traced_instancemethod_delegator
    return proxy(method, args, keywords)
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/autologging.py", line 1196, in __call__
    value = function(*args, **keywords)
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/fsspec/fuse.py", line 25, in getattr
    raise FuseOSError(ENOENT)
fuse.FuseOSError: [Errno 2] No such file or directory
TRACE    fsspec.fuse.FUSEr:fuse.py:70 CALL *('/test', 33188) **{}
TRACE    fsspec.fuse.FUSEr:fuse.py:70 RETURN 0
TRACE    fsspec.fuse.FUSEr:fuse.py:20 CALL *('/test', 0) **{}
TRACE    fsspec.fuse.FUSEr:fuse.py:20 RETURN {'st_uid': 1000, 'st_gid': 1000, 'st_mode': 33279, 'st_size': 0, 'st_blksize': 5242880, 'st_nlink': 1, 'st_atime': 1568647084.8037694, 'st_ctime': 1568647084.8037698, 'st_mtime': 1568647084.80377}
TRACE    fsspec.fuse.FUSEr:fuse.py:116 CALL *('/test', 21505, 140734676863040, 0, 0, None) **{}
DEBUG    fuse:fuse.py:740 FUSE operation ioctl raised a <class 'fuse.FuseOSError'>, returning errno 25.
Traceback (most recent call last):
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/fuse.py", line 734, in _wrapper
    return func(*args, **kwargs) or 0
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/fuse.py", line 1061, in ioctl
    cmd, arg, fh, flags, data)
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/fuse.py", line 1076, in __call__
    return getattr(self, op)(*args)
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/autologging.py", line 1039, in autologging_traced_instancemethod_delegator
    return proxy(method, args, keywords)
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/autologging.py", line 1196, in __call__
    value = function(*args, **keywords)
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/fsspec/fuse.py", line 117, in ioctl
    super().ioctl(path, cmd, arg, fip, flags, data)
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/fuse.py", line 1143, in ioctl
    raise FuseOSError(errno.ENOTTY)
fuse.FuseOSError: [Errno 25] Inappropriate ioctl for device
DEBUG    fuse:fuse.py:740 FUSE operation getxattr raised a <class 'fuse.FuseOSError'>, returning errno 95.
Traceback (most recent call last):
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/fuse.py", line 734, in _wrapper
    return func(*args, **kwargs) or 0
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/fuse.py", line 910, in getxattr
    name.decode(self.encoding), *args)
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/fuse.py", line 1076, in __call__
    return getattr(self, op)(*args)
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/fuse.py", line 1131, in getxattr
    raise FuseOSError(ENOTSUP)
fuse.FuseOSError: [Errno 95] Operation not supported
TRACE    fsspec.fuse.FUSEr:fuse.py:65 CALL *('/test', b'data', 0, 0) **{}
TRACE    fsspec.fuse.FUSEr:fuse.py:65 RETURN 4
DEBUG    fsspec.tests.test_fuse:test_fuse.py:39 info
TRACE    fsspec.fuse.FUSEr:fuse.py:103 CALL *('/test', 0) **{}
DEBUG    fsspec.tests.test_fuse:test_fuse.py:42 open
TRACE    fsspec.fuse.FUSEr:fuse.py:103 RETURN 0
TRACE    fsspec.fuse.FUSEr:fuse.py:77 CALL *('/test', 32768) **{}
TRACE    fsspec.fuse.FUSEr:fuse.py:77 RETURN 1
TRACE    fsspec.fuse.FUSEr:fuse.py:20 CALL *('/test', None) **{}
TRACE    fsspec.fuse.FUSEr:fuse.py:20 RETURN {'st_uid': 1000, 'st_gid': 1000, 'st_mode': 33279, 'st_size': 4, 'st_blksize': 5242880, 'st_nlink': 1, 'st_atime': 1568647084.805991, 'st_ctime': 1568647084.8059912, 'st_mtime': 1568647084.8059914}
TRACE    fsspec.fuse.FUSEr:fuse.py:116 CALL *('/test', 21505, 140734676863040, 1, 0, None) **{}
DEBUG    fuse:fuse.py:740 FUSE operation ioctl raised a <class 'fuse.FuseOSError'>, returning errno 25.
Traceback (most recent call last):
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/fuse.py", line 734, in _wrapper
    return func(*args, **kwargs) or 0
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/fuse.py", line 1061, in ioctl
    cmd, arg, fh, flags, data)
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/fuse.py", line 1076, in __call__
    return getattr(self, op)(*args)
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/autologging.py", line 1039, in autologging_traced_instancemethod_delegator
    return proxy(method, args, keywords)
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/autologging.py", line 1196, in __call__
    value = function(*args, **keywords)
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/fsspec/fuse.py", line 117, in ioctl
    super().ioctl(path, cmd, arg, fip, flags, data)
  File "/work/home/lexaf/workspace/asford/filesystem_spec/filesystem_spec/.tox/fuse/lib/python3.7/site-packages/fuse.py", line 1143, in ioctl
    raise FuseOSError(errno.ENOTTY)
fuse.FuseOSError: [Errno 25] Inappropriate ioctl for device
TRACE    fsspec.fuse.FUSEr:fuse.py:116 CALL *('/test', 21505, 140734676862928, 1, 0, None) **{}
DEBUG    fsspec.tests.test_fuse:test_fuse.py:45 read
------------------------------------------------------------------ Captured stdout teardown ------------------------------------------------------------------
```

asford commented:
Confirmed that this issue reproduces [on travis](https://travis-ci.org/intake/filesystem_spec/jobs/585643869#L689-L710). Would you mind running the test locally?

martindurant commented on fsspec/fuse.py:
```diff
@@ -112,6 +119,8 @@ def release(self, path, fh):
         raise NotImplementedError
 
+    def ioctl(self, path, cmd, arg, fip, flags, data):
+        super().ioctl(path, cmd, arg, fip, flags, data)
```
Why is this ever called? I don't think it makes sense to try to do anything "device"-oriented for fsspec implementations. The line of code here, is simply a no-op, though, right?

asford replied:
Yes, this is just a shim in place while I diagnose the fuse test error. Will drop, unless it appears necessary for the test pass, before merge.

martindurant commented on fsspec/tests/test_fuse.py:
```diff
@@ -5,37 +5,59 @@
 import time
 
+import logging
+logger = logging.getLogger(__name__)
```
The logger is only for CI debugging purposes?

asford replied:
Yes

martindurant commented:
test_fuse currently passes for me, with py 3.7.3, fusepy 3.0.1

asford commented:
Roger, I'm still seeing the error locally and have pushed version pins to Travis to debug there. Would you mind running the tox configuration in your environment to help determine if the error was introduced by this change?

asford commented:
Can you `tox -e fuse`? This will setup and run the precise repro. It's not installed in `dev` as the error causes test runs to deadlock in a non-`kill -9` responsive state in my environment.

martindurant commented:
Yep, fails (read-only-filesystem, file-not-found, etc) and hangs.

asford commented:
Well, shoot, that's baffling.

How are you configuring your local environment? I'll try to get an equivalent version running locally.

I'm getting the test failure with a manually created environment containing `python 3.7.4`, `fusepy 3.0.1` running against `intake/master`:

```bash
+~/workspace/asford/filesystem_spec/filesystem_spec$ which python
/work/home/lexaf/workspace/asford/filesystem_spec/.conda/bin/python

+~/workspace/asford/filesystem_spec/filesystem_spec$ git show HEAD
commit 65d36f2719e493bbc6e201f1cdaf3ac2af647327 (HEAD, tag: 0.4.5, intake/master)
Merge: ff9e15b a45bd9d
Author: Martin Durant <martindurant@users.noreply.github.com>
Date:   Mon Sep 16 12:10:38 2019 -0400

    Merge pull request #132 from martindurant/WFCache

    whole-file cache

+~/workspace/asford/filesystem_spec/filesystem_spec$ conda list
# packages in environment at /work/home/lexaf/workspace/asford/filesystem_spec/.conda:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main
atomicwrites              1.3.0                    py37_1
attrs                     19.1.0                   py37_1
backcall                  0.1.0                    py37_0
ca-certificates           2019.5.15                     1
certifi                   2019.6.16                py37_1
decorator                 4.4.0                    py37_1
fusepy                    3.0.1                      py_0    conda-forge
importlib_metadata        0.19                     py37_0
ipython                   7.8.0            py37h39e3cac_0
ipython_genutils          0.2.0                    py37_0
jedi                      0.15.1                   py37_0
libedit                   3.1.20181209         hc058e9b_0
libffi                    3.2.1                hd88cf55_4
libgcc-ng                 9.1.0                hdf63c60_0
libstdcxx-ng              9.1.0                hdf63c60_0
lz4                       2.1.9            py37h7b6447c_0
lz4-c                     1.8.1.2              h14c3975_0
more-itertools            7.2.0                    py37_0
ncurses                   6.1                  he6710b0_1
openssl                   1.1.1d               h7b6447c_1
packaging                 19.1                     py37_0
parso                     0.5.1                      py_0
pexpect                   4.7.0                    py37_0
pickleshare               0.7.5                    py37_0
pip                       19.2.2                   py37_0
pluggy                    0.12.0                     py_0
prompt_toolkit            2.0.9                    py37_0
ptyprocess                0.6.0                    py37_0
py                        1.8.0                    py37_0
pygments                  2.4.2                      py_0
pyparsing                 2.4.2                      py_0
pytest                    5.0.1                    py37_0
pytest-instafail          0.4.1                      py_0    conda-forge
python                    3.7.4                h265db76_1
python-snappy             0.5.4            py37he6710b0_0
readline                  7.0                  h7b6447c_5
setuptools                41.0.1                   py37_0
six                       1.12.0                   py37_0
snappy                    1.1.7                hbae5bb6_3
sqlite                    3.29.0               h7b6447c_0
tk                        8.6.8                hbc83047_0
traitlets                 4.3.2                    py37_0
wcwidth                   0.1.7                    py37_0
wheel                     0.33.4                   py37_0
xz                        5.2.4                h14c3975_4
zipp                      0.5.2                      py_0
zlib                      1.2.11               h7b6447c_3
zstandard                 0.12.0           py37he1b5a44_0    conda-forge

+~/workspace/asford/filesystem_spec/filesystem_spec$ apt list fuse
Listing... Done
fuse/xenial-updates,now 2.9.4-1ubuntu3.1 amd64 [installed]
N: There is 1 additional version. Please use the '-a' switch to see it
```

martindurant commented:
Full package listing follows. This is running on OSX, with FUSE for macOS 3.7.1 . I notice that the mount point was not cleaned up when the test finished.

```
alabaster                 0.7.12                     py_0    conda-forge
appdirs                   1.4.3                      py_1    conda-forge
appnope                   0.1.0                 py37_1000    conda-forge
arrow-cpp                 0.14.1           py37h43d7656_0    conda-forge
asciitree                 0.3.3                      py_2    conda-forge
asn1crypto                0.24.0                py37_1003    conda-forge
astropy                   3.2.1            py37h01d97ff_0    conda-forge
atomicwrites              1.3.0                      py_0    conda-forge
attrs                     19.1.0                     py_0    conda-forge
aws-sam-translator        1.13.0                     py_0    conda-forge
aws-xray-sdk              0.95                       py_0    conda-forge
babel                     2.7.0                      py_0    conda-forge
backcall                  0.1.0                      py_0    conda-forge
backports                 1.0                        py_2    conda-forge
backports.tempfile        1.0                        py_0    conda-forge
backports.weakref         1.0.post1             py37_1000    conda-forge
bcrypt                    3.1.6            py37h01d97ff_1    conda-forge
black                     19.3b0                     py_0    conda-forge
bleach                    3.1.0                      py_0    conda-forge
blinker                   1.4                        py_1    conda-forge
bokeh                     1.3.4                    py37_0    conda-forge
boost-cpp                 1.70.0               h75728bb_2    conda-forge
boto                      2.49.0                     py_0    conda-forge
boto3                     1.9.197                    py_0    conda-forge
botocore                  1.12.197                   py_0    conda-forge
brotli                    1.0.7             h6de7cb9_1000    conda-forge
bzip2                     1.0.8                h01d97ff_0    conda-forge
c-ares                    1.15.0            h01d97ff_1001    conda-forge
ca-certificates           2019.6.16            hecc5488_0    conda-forge
cachetools                2.1.0                      py_0    conda-forge
cartopy                   0.17.0          py37hd103db2_1006    conda-forge
certifi                   2019.6.16                py37_1    conda-forge
cffi                      1.12.3           py37hccf1714_0    conda-forge
cfn-lint                  0.22.0                     py_0    conda-forge
cftime                    1.0.3.4         py37h1e5eb4f_1001    conda-forge
chardet                   3.0.4                 py37_1003    conda-forge
click                     7.0                        py_0    conda-forge
cloudpickle               1.2.1                      py_0    conda-forge
cmarkgfm                  0.4.2            py37h1de35cc_2    conda-forge
colorcet                  2.0.1                      py_0    conda-forge
cookies                   2.2.1                      py_0    conda-forge
coverage                  4.5.4            py37h01d97ff_0    conda-forge
cryptography              2.7              py37h212c5bf_0    conda-forge
curl                      7.65.3               h22ea746_0    conda-forge
cycler                    0.10.0                     py_1    conda-forge
cython                    0.29.13          py37h6de7cb9_0    conda-forge
cytoolz                   0.10.0           py37h01d97ff_0    conda-forge
dask                      2.2.0+38.g266c314           dev_0    <develop>
dask-sphinx-theme         1.1.1                      py_0    conda-forge
datashader                0.7.0                      py_0    conda-forge
datashape                 0.5.4                      py_1    conda-forge
dbus                      1.13.6               h2f22bb5_0    conda-forge
decorator                 4.4.0                      py_0    conda-forge
defusedxml                0.5.0                      py_1    conda-forge
dfviz                     0.1.0+1.g59d40b4.dirty           dev_0    <develop>
distributed               2.2.0                      py_1    conda-forge
docker-py                 4.0.2                    py37_0    conda-forge
docker-pycreds            0.4.0                      py_0    conda-forge
docutils                  0.14                  py37_1001    conda-forge
double-conversion         3.1.5                h6de7cb9_1    conda-forge
ecdsa                     0.13                       py_0    conda-forge
entrypoints               0.3                   py37_1000    conda-forge
expat                     2.2.5             h6de7cb9_1003    conda-forge
fastavro                  0.22.3           py37h01d97ff_0    conda-forge
fasteners                 0.14.1                     py_3    conda-forge
fastparquet               0.3.2            py37heacc8b8_0    conda-forge
flake8                    3.7.8                    py37_1    conda-forge
flask                     1.1.1                      py_1    conda-forge
freetype                  2.10.0               h24853df_0    conda-forge
fsspec                    0.4.4+27.g82d90b4           dev_0    <develop>
fusepy                    3.0.1                      py_0    conda-forge
future                    0.17.1                py37_1000    conda-forge
gcsfs                     0.3.0                      py_0    conda-forge
geos                      3.7.2                h6de7cb9_1    conda-forge
gettext                   0.19.8.1          h46ab8bc_1002    conda-forge
gflags                    2.2.2             h6de7cb9_1001    conda-forge
glib                      2.58.3            h9d45998_1002    conda-forge
glog                      0.4.0                h6de7cb9_1    conda-forge
gmp                       6.1.2             h0a44026_1000    conda-forge
google-auth               1.6.3                      py_0    conda-forge
google-auth-oauthlib      0.4.0                      py_0    conda-forge
grpc-cpp                  1.22.0               h5c95ec2_0    conda-forge
hadoop-test-cluster       0.1.0                    pypi_0    pypi
hdf4                      4.2.13            hf3c6af0_1002    conda-forge
hdf5                      1.10.5          nompi_h0cbb7df_1100    conda-forge
heapdict                  1.0.0                 py37_1000    conda-forge
holoviews                 1.12.1                     py_2    conda-forge
hvplot                    0.4.0                      py_1    conda-forge
icu                       64.2                 h6de7cb9_0    conda-forge
idna                      2.8                   py37_1000    conda-forge
imageio                   2.5.0                    py37_0    conda-forge
imagesize                 1.1.0                      py_0    conda-forge
importlib_metadata        0.18                     py37_0    conda-forge
intake                    0.5.3                     dev_0    <develop>
intake-parquet            0.2.2                     dev_0    <develop>
intake-xarray             0.3.1                      py_0    conda-forge
ipykernel                 5.1.1            py37h5ca1d4c_0    conda-forge
ipython                   7.7.0            py37h5ca1d4c_0    conda-forge
ipython_genutils          0.2.0                      py_1    conda-forge
ipywidgets                7.5.1                      py_0    conda-forge
itsdangerous              1.1.0                      py_0    conda-forge
jedi                      0.14.1                   py37_0    conda-forge
jinja2                    2.10.1                     py_0    conda-forge
jmespath                  0.9.4                      py_0    conda-forge
jpeg                      9c                h1de35cc_1001    conda-forge
jsondiff                  1.1.2                      py_0    conda-forge
jsonpatch                 1.24                       py_0    conda-forge
jsonpickle                1.2                        py_0    conda-forge
jsonpointer               2.0                        py_0    conda-forge
jsonschema                3.0.0a3               py37_1000    conda-forge
jupyter                   1.0.0                      py_2    conda-forge
jupyter_client            5.3.1                      py_0    conda-forge
jupyter_console           6.0.0                      py_0    conda-forge
jupyter_core              4.4.0                      py_0    conda-forge
kiwisolver                1.1.0            py37h770b8ee_0    conda-forge
krb5                      1.16.3            hcfa6398_1001    conda-forge
libblas                   3.8.0               11_openblas    conda-forge
libcblas                  3.8.0               11_openblas    conda-forge
libcurl                   7.65.3               h16faf7d_0    conda-forge
libcxx                    8.0.0                         4    conda-forge
libcxxabi                 8.0.0                         4    conda-forge
libedit                   3.1.20170329      hcfe32e1_1001    conda-forge
libevent                  2.1.10               h212c5bf_0    conda-forge
libffi                    3.2.1             h6de7cb9_1006    conda-forge
libgfortran               3.0.1                         0    conda-forge
libiconv                  1.15              h01d97ff_1005    conda-forge
liblapack                 3.8.0               11_openblas    conda-forge
libnetcdf                 4.6.2             h1a02027_1002    conda-forge
libopenblas               0.3.6                hd44dcd8_6    conda-forge
libpng                    1.6.37               h2573ce8_0    conda-forge
libprotobuf               3.8.0                hfbae3c0_0    conda-forge
libsodium                 1.0.17               h01d97ff_0    conda-forge
libssh2                   1.8.2                hcdc9a53_2    conda-forge
libtiff                   4.0.10            hd08fb8f_1003    conda-forge
line_profiler             2.1.2           py37h01d97ff_1003    conda-forge
llvmlite                  0.29.0           py37h3fea490_1    conda-forge
locket                    0.2.0                      py_2    conda-forge
lz4-c                     1.8.3             h6de7cb9_1001    conda-forge
lzo                       2.10              h1de35cc_1000    conda-forge
markdown                  3.1.1                      py_0    conda-forge
markupsafe                1.1.1            py37h1de35cc_0    conda-forge
matplotlib                3.1.1                    py37_1    conda-forge
matplotlib-base           3.1.1            py37h3a684a6_1    conda-forge
mccabe                    0.6.1                      py_1    conda-forge
metpy                     0.10.2                   py37_0    conda-forge
mistune                   0.8.4           py37h1de35cc_1000    conda-forge
mock                      3.0.5                    py37_0    conda-forge
monotonic                 1.5                        py_0    conda-forge
more-itertools            7.2.0                      py_0    conda-forge
moto                      1.3.12                     py_2    conda-forge
msgpack-numpy             0.4.4.3                    py_0    conda-forge
msgpack-python            0.6.1            py37h04f5b5a_0    conda-forge
multipledispatch          0.6.0                      py_0    conda-forge
nbconvert                 5.5.0                      py_0    conda-forge
nbformat                  4.4.0                      py_1    conda-forge
ncurses                   6.1               h0a44026_1002    conda-forge
netcdf4                   1.5.1.2          py37h876e625_1    conda-forge
networkx                  2.3                        py_0    conda-forge
notebook                  6.0.0                    py37_0    conda-forge
numba                     0.45.1           py37h86efe34_0    conda-forge
numcodecs                 0.6.3            py37h0a44026_0    conda-forge
numpy                     1.16.4           py37h6b0580a_0    conda-forge
numpydoc                  0.8.0                    py37_0
oauthlib                  3.0.1                      py_0    conda-forge
olefile                   0.46                       py_0    conda-forge
openssl                   1.1.1c               h01d97ff_0    conda-forge
owslib                    0.18.0                     py_0    conda-forge
packaging                 19.0                       py_0    conda-forge
pandas                    0.25.0           py37h86efe34_0    conda-forge
pandoc                    2.7.3                         0    conda-forge
pandocfilters             1.4.2                      py_1    conda-forge
panel                     0.6.2                         0    conda-forge
param                     1.9.1                      py_0    conda-forge
paramiko                  2.6.0                    py37_0    conda-forge
parquet-cpp               1.5.1                         2    conda-forge
parso                     0.5.1                      py_0    conda-forge
partd                     1.0.0                      py_0    conda-forge
pcre                      8.41              h0a44026_1003    conda-forge
pexpect                   4.7.0                    py37_0    conda-forge
pickleshare               0.7.5                 py37_1000    conda-forge
pillow                    6.1.0            py37h75ffe9a_1    conda-forge
pint                      0.9                      py37_2    conda-forge
pip                       19.2.1                   py37_0    conda-forge
pkginfo                   1.5.0.1                    py_0    conda-forge
pluggy                    0.12.0                     py_0    conda-forge
pooch                     0.5.2                    py37_0    conda-forge
proj4                     5.2.0             h6de7cb9_1004    conda-forge
prometheus_client         0.7.1                      py_0    conda-forge
prompt_toolkit            2.0.9                      py_0    conda-forge
psutil                    5.6.3            py37h01d97ff_0    conda-forge
ptyprocess                0.6.0                   py_1001    conda-forge
py                        1.8.0                      py_0    conda-forge
pyarrow                   0.14.1           py37he1943e6_0    conda-forge
pyasn1                    0.4.6                      py_0    conda-forge
pyasn1-modules            0.2.6                      py_0    conda-forge
pycodestyle               2.5.0                      py_0    conda-forge
pycparser                 2.19                     py37_1    conda-forge
pycryptodome              3.8.2            py37h44b798e_0    conda-forge
pyct                      0.4.6                      py_0    conda-forge
pyct-core                 0.4.6                      py_0    conda-forge
pyepsg                    0.4.0                      py_0    conda-forge
pyflakes                  2.1.1                      py_0    conda-forge
pyftpdlib                 1.5.5                      py_0    conda-forge
pygments                  2.4.2                      py_0    conda-forge
pyjwt                     1.7.1                      py_0    conda-forge
pykdtree                  1.3.1           py37h917ab60_1002    conda-forge
pynacl                    1.3.0           py37h1de35cc_1000    conda-forge
pyopenssl                 19.0.0                   py37_0    conda-forge
pyparsing                 2.4.2                      py_0    conda-forge
pyproj                    1.9.6           py37h01d97ff_1002    conda-forge
pyqt                      5.9.2            py37h2a560b1_2    conda-forge
pyrsistent                0.15.4           py37h01d97ff_0    conda-forge
pyshp                     2.1.0                      py_0    conda-forge
pysocks                   1.7.0                    py37_0    conda-forge
pytest                    5.0.1                    py37_1    conda-forge
pytest-arraydiff          0.3                        py_0    conda-forge
pytest-astropy            0.5.0                      py_0    conda-forge
pytest-cov                2.7.1                      py_0    conda-forge
pytest-doctestplus        0.3.0                      py_0    conda-forge
pytest-openfiles          0.4.0                      py_0    conda-forge
pytest-remotedata         0.3.1                      py_0    conda-forge
python                    3.7.3                h93065d6_1    conda-forge
python-dateutil           2.8.0                      py_0    conda-forge
python-jose               2.0.2                      py_0    conda-forge
python-snappy             0.5.4            py37h1e06ddd_0    conda-forge
pytz                      2019.2                     py_0    conda-forge
pyviz_comms               0.7.2                      py_0    conda-forge
pywavelets                1.0.3            py37h1e5eb4f_1    conda-forge
pyyaml                    5.1.2            py37h01d97ff_0    conda-forge
pyzmq                     18.0.2           py37hee98d25_2    conda-forge
qt                        5.9.7                h8cf7e54_3    conda-forge
qtconsole                 4.5.2                      py_0    conda-forge
re2                       2019.08.01           h6de7cb9_0    conda-forge
readline                  8.0                  hcfe32e1_0    conda-forge
readme_renderer           24.0                       py_0    conda-forge
requests                  2.22.0                   py37_1    conda-forge
requests-oauthlib         1.2.0                      py_0    conda-forge
requests-toolbelt         0.9.1                      py_0    conda-forge
responses                 0.9.0                      py_0    conda-forge
rsa                       3.4.2                      py_1    conda-forge
ruamel                    1.0                      py37_0    conda-forge
ruamel.yaml               0.16.0           py37h01d97ff_0    conda-forge
s3fs                      0.3.3                     dev_0    <develop>
s3transfer                0.2.1                    py37_0    conda-forge
scikit-image              0.15.0           py37h86efe34_2    conda-forge
scipy                     1.3.0            py37hab3da7d_1    conda-forge
send2trash                1.5.0                      py_0    conda-forge
setuptools                41.0.1                   py37_0    conda-forge
shapely                   1.6.4           py37h0567c5e_1006    conda-forge
sip                       4.19.8          py37h0a44026_1000    conda-forge
six                       1.12.0                py37_1000    conda-forge
snakeviz                  2.0.0                      py_0    conda-forge
snappy                    1.1.7             h6de7cb9_1002    conda-forge
snowballstemmer           1.9.0                      py_0    conda-forge
sortedcontainers          2.1.0                      py_0    conda-forge
sphinx                    2.0.1                      py_0    conda-forge
sphinx-click              2.1.0                      py_0    conda-forge
sphinx_rtd_theme          0.4.3                      py_0    conda-forge
sphinxcontrib-applehelp   1.0.1                      py_0    conda-forge
sphinxcontrib-devhelp     1.0.1                      py_0    conda-forge
sphinxcontrib-htmlhelp    1.0.2                      py_0    conda-forge
sphinxcontrib-jsmath      1.0.1                      py_0    conda-forge
sphinxcontrib-qthelp      1.0.2                      py_0    conda-forge
sphinxcontrib-serializinghtml 1.1.1                      py_0    conda-forge
sqlalchemy                1.3.6            py37h01d97ff_0    conda-forge
sqlite                    3.29.0               hb7d70f7_0    conda-forge
sshpubkeys                3.1.0                      py_0    conda-forge
tblib                     1.4.0                      py_0    conda-forge
terminado                 0.8.2                    py37_0    conda-forge
testpath                  0.4.2                   py_1001    conda-forge
thrift                    0.11.0          py37h0a44026_1001    conda-forge
thrift-cpp                0.12.0            h874095a_1004    conda-forge
tk                        8.6.9             h2573ce8_1002    conda-forge
toml                      0.10.0                     py_0    conda-forge
toolz                     0.10.0                     py_0    conda-forge
tornado                   6.0.3            py37h01d97ff_0    conda-forge
tqdm                      4.32.2                     py_0    conda-forge
traitlets                 4.3.2                 py37_1000    conda-forge
twine                     1.13.0                     py_1    conda-forge
uriparser                 0.9.3                h6de7cb9_1    conda-forge
urllib3                   1.25.3                   py37_0    conda-forge
us_crime                  0.1.0                         2    intake
vcrpy                     2.1.0                    py37_0    conda-forge
wcwidth                   0.1.7                      py_1    conda-forge
webencodings              0.5.1                      py_1    conda-forge
websocket-client          0.56.0                   py37_0    conda-forge
werkzeug                  0.15.5                     py_0    conda-forge
wheel                     0.33.4                   py37_0    conda-forge
widgetsnbextension        3.5.1                    py37_0    conda-forge
wrapt                     1.11.2           py37h01d97ff_0    conda-forge
xarray                    0.12.3                     py_0    conda-forge
xmltodict                 0.12.0                     py_0    conda-forge
xrviz                     0.1.1+38.g9eeadb3           dev_0    <develop>
xz                        5.2.4             h1de35cc_1001    conda-forge
yaml                      0.1.7             h1de35cc_1001    conda-forge
zarr                      2.3.2                    py37_0    conda-forge
zeromq                    4.3.2                h6de7cb9_2    conda-forge
zict                      1.0.0                      py_0    conda-forge
zipp                      0.5.2                      py_0    conda-forge
zlib                      1.2.11            h01d97ff_1005    conda-forge
zstd                      1.4.0                ha9f0a20_0    conda-forge
```

asford commented:
Interesting. I suspect the hung failure issue is due to some from of intra-thread deadlock when an error is raised in the fuse-hosted thread while handling a system call originating in a different thread. Given that this is an ...unorthodox... use case for fuse I've rewritten the test to launch the fuse component in a subprocess.

Bad news first, it still fails. Good news second, it's failing with an meaningful exception and exits cleanly. Looks like the `remove` function is attempting getattr on a non-existing fuse related file? [`FileNotFoundError: /mounted/.fuse_hidden0000000200000001`](https://travis-ci.org/intake/filesystem_spec/jobs/585781047#L743-L780) Does this ring any bells in your previous work with fuse?

asford commented:
Ok, this is ready for review. I've fixed the py35 incompatibilities I introduced in #130 and have fixed the fuse test so it runs properly in the tox/linux environment. I've added skip operations for the expected zipfile-based failures in py35, but there are several remaining py35 failures. The remaining failures reproduce in current master running under py35, [`test_memory.py::test_ls`](https://travis-ci.org/intake/filesystem_spec/jobs/585801339#L689) and [`test_zip.py::test_pickle`](https://travis-ci.org/intake/filesystem_spec/jobs/585801339#L708), would you mind investigating on your end as a follow-up to this PR?

asford commented:
Ack, found a few more py35 syntax errors introduced by #137. Fixed here via e150720.

I do think some form of py35 test support in CI would be valuable. Perhaps marking the current py35 failures exposed by this PR as xfail on py35 so we can get this green and merge?

asford commented:
@martindurant Friendly ping on this review.

 If the tox-based testing configuration is a problem I'd be happy to split the PR apart, but would like to resolve and provide test coverage for the py35 incompatibilities and fuse test issues addressed here.

martindurant commented:
Yep, best xfail them and circle back.
I assume the test_ls has to do with ordering of output (should probably be explicitly sorted), but test_pickle may not be trivially fixable.

asford commented:
Great, py35 failures have been marked xfail and CI is green. This ready to merge.
